### PR TITLE
Add mathematical constants

### LINKED ISSUE #4507: Mathematical constants and functions in Kokkos
Hi all,

What is the best way to use mathematical constants such as pi in a portable way alongside kokkos? Does Kokkos provide an interface for it and is it guaranteed to have in all future releases too? I am afraid to use things like M_PI from cmath because of portability. Similarly, for mathematical functions such as sin, cos, erf etc. currently I use std::cos, std::sin, std::erf from cmath (inside Kokkos loops also). Is this fine or is there a better way of doing it from Kokkos?   

Thanks

### PR DESCRIPTION
**Pull request by dalg24**
Fix #4507
Backporting C++20 facility https://eel.is/c++draft/numbers
Individual traits with a `value` member for each 13 constants defined in the standard.

### CODE REVIEW

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
Hold on before you merge that one.  I probably should have used variable templates.  I forgot they were available in C++14

**Comment by dalg24** [comment]
Except that inline variables only come with C++17...

**Review by fnrizzi — APPROVED** [review]

**Review by crtrott — CHANGES_REQUESTED** [review]
Discussed this with @dalg24 today there is a slightly different implementation idea in order to avoid introducing traits which are not part of the standard in the public Kokkos namespace (i.e. put the trait into Impl, and only the variable declaration in Kokkos)

File: core/unit_test/TestMathematicalConstants.hpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+
+template <class T>
+KOKKOS_FUNCTION T *take_address_of(T &arg) {
```
Comment (nliber) [inline core/unit_test/TestMathematicalConstants.hpp:49]: Just curious:: why not use `std::addressof`?

**Review by nliber — APPROVED** [review]

File: core/unit_test/TestMathematicalConstants.hpp (hunk 1 above)
Comment (dalg24) [reply core/unit_test/TestMathematicalConstants.hpp:49]: `__host__ __device__` annotation missing

Comment (masterleinad) [reply core/unit_test/TestMathematicalConstants.hpp:49]: https://godbolt.org/z/dhW9zzTE5

**Review by crtrott — APPROVED** [review]

**Pull request merged**